File: 24.swap-nodes-in-pairs.java
/*
 * @lc app=leetcode id=24 lang=java
 *
 * [24] Swap Nodes in Pairs
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        if (head == null) {
            return head;
        }
        if (head.next == null) {
            return head;
        }
        ListNode dummy = new ListNode(-1, head);
        ListNode iter = dummy;
        while (iter != null && iter.next != null && iter.next.next != null) {
            ListNode p = iter.next;
            ListNode q = iter.next.next;
            ListNode t = iter.next.next.next;
            iter.next = q;
            q.next = p;
            p.next = t;
            iter = iter.next.next;
        }
        return dummy.next;
    }
}
// @lc code=end

File: 24.swap-nodes-in-pairs.cpp
/*
 * @lc app=leetcode id=24 lang=cpp
 *
 * [24] Swap Nodes in Pairs
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if (head == nullptr || head->next == nullptr) {
            return head;
        }
        ListNode l {-1, head};
        ListNode *dummy = &l;
        ListNode *prev = dummy->next, *curr = dummy->next->next;
        while (curr != nullptr) {
            std::swap(prev->val, curr->val);
            prev = curr->next;
            curr = (curr->next == nullptr) ? nullptr : curr->next->next;
        }
        return dummy->next;
    }
};
// @lc code=end
